Support new emoji and fallback from blob to Google for new emoji.

evykassirer opened the pull request:
Fixes #19371 and #21037.

- [x] **Do not merge until mobile can support this! See #22476**

This PR:

* Adds new emoji added since 2017, using their CLDR names and aliases.
* Supports falling back to Google Modern for emoji not supported by Google Classic.
* Updates the settings menu to note that Google Classic is no longer supported, linking to a new helpdesk blurb.
* Creates an API for mobile to see what emoji are currently supported by the client.

![image](https://user-images.githubusercontent.com/5634097/168696413-b3e4e99d-eaf4-4d7d-8cef-eb472990a3dc.png)

<img width="424" alt="image" src="https://user-images.githubusercontent.com/5634097/177417552-8b4397a3-091e-46f2-99ef-b53bf8f44a65.png">

<img width="521" alt="image" src="https://user-images.githubusercontent.com/5634097/177417580-7150dd0d-64b2-445b-bda3-46ca5c8d2da0.png">

timabbott commented:
For the `emoji_names.py` part, would it be cleaner to import CLDR or similar as a dependency, and just have it fill in any value not present in `emoji_names.py` already?

evykassirer commented:
It takes less space to just have this instead of all the CLDR package, though I agree that the way I manually did it is a bit jank. If we choose to move fully to CLDR there'll be a cleaner way to do this, so I'm sort of waiting to see how much we want to tweak names vs just use CLDR, and planning to clean this process up more after that. Does that sound ok to you?

evykassirer commented on zerver/models.py:
```diff
@@ -1495,9 +1495,9 @@ class UserBaseSettings(models.Model):
-        (GOOGLE_BLOB_EMOJISET, "Google classic"),
         (TWITTER_EMOJISET, "Twitter"),
         (TEXT_EMOJISET, "Plain text"),
+        (GOOGLE_BLOB_EMOJISET, f"Google classic ({_('deprecated')})"),
```
Maybe it's better to add the "deprecated" text directly in the settings menu like the tooltip, and that would also make it easier to make it a link.

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -1577,4 +1577,1910 @@
+        "canonical_name": "woman teacher ",
+        "aliases": ["instructor", "professor", "teacher", "woman"],
+    },
+    "1f469-200d-1f3ed": {
+        "canonical_name": "woman factory worker ",
+        "aliases": ["assembly", "factory", "industrial", "woman", "worker"],
+    },
+    "1f469-200d-1f466-200d-1f466": {
+        "canonical_name": "family: woman, boy, boy ",
+        "aliases": ["boy", "family", "woman"],
```
In our system, a given name maps to a single emoji. A `canonical_name` or an alias for any given emoji becomes a name for that emoji, and can only point to one of them.

These terms that are populating `aliases` here aren't meant in CLDR to be aliases in that sense -- they're more like search keywords, where the same keyword can naturally be present on many items. Some discussion here: https://github.com/zulip/zulip/issues/18121#issuecomment-1110008908

Pending #18121, we'll want a story for what to do with CLDR's keywords when importing their emoji names. It should avoid creating duplicate aliases.

A very simple thing to do would be to drop the keywords in this CLDR import. But for many emoji, they provide useful synonyms or alternative cultural interpretations.

A still pretty simple rule that might get us a long way is to drop any keywords that also appear in the canonical name (in particular as whole words, with no letters before or after). If the user searches with those keywords in a way that would match the aliases, then it would also match the name directly.

(See https://github.com/zulip/zulip/commit/61b12dc25c4c260041774279111ad1b7a461209b / https://github.com/zulip/zulip/pull/21871 for recent changes to emoji search. It still won't get everything one might want -- in the examples above, I believe it'd currently work for "factory worker" but not for "boy boy" (tripped up by punctuation) or for "woman worker" or "worker factory" (won't match out of order.) But the aliases wouldn't help, because we don't try to spread a query across multiple names for an emoji anyway.)

Then after that… maybe just drop any aliases that are still duplicates. We could probably do something cleverer than that, but I think effort in that direction is probably best spent working toward making the search compatible with using CLDR-style keywords instead of aliases, as part of moving toward #18121.

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -1577,4 +1577,1910 @@
     "1f6a9": {"canonical_name": "triangular_flag", "aliases": []},
     # solidarity from iemoji
     "1f38c": {"canonical_name": "crossed_flags", "aliases": ["solidarity"]},
+    "00a9": {"canonical_name": "copyright ", "aliases": ["C", "copyright"]},
```
It'd probably also be helpful to write down somewhere (even just in a commit message) whatever script or process you have for importing the CLDR names and keywords into this form. That'll simplify rerunning it as we adjust things about how it works.

evykassirer replied:
Yup I accidentally added a trailing space -- oops! Thanks :)

I'm making a script to check in that generates `emoji_names.py` from CLDR data and hopefully that helps us in the future :)

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -1577,4 +1577,1910 @@
+    "1f94f": {"canonical_name": "flying disc ", "aliases": ["flying disc", "ultimate"]},
+    "1f95f": {
+        "canonical_name": "dumpling ",
+        "aliases": ["dumpling", "empanada", "gyōza", "jiaozi", "pierogi", "potsticker"],
```
Hmm, I wonder if our emoji search tries a diacritics-removed match, so that "gyoza" would match this. I guess it should!

Similarly "piña colada" below, and a few of the territory names like "Côte d’Ivoire".

evykassirer replied:
Seems like it didn't! I manually added some aliases, but that's not the most ideal approach.

gnprice replied:
Discussion here: https://chat.zulip.org/#narrow/stream/6-frontend/topic/update.20emoji.20.2319371/near/1392206

gnprice reviewed (commented):
Thanks @evykassirer! Looking forward to seeing this in.

gnprice commented:
(Chat thread is here, for cross-reference: https://chat.zulip.org/#narrow/stream/6-frontend/topic/update.20emoji.20.2319371/near/1391533 )

gnprice commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -0,0 +1,141 @@
+]
+
+
+def main() -> None:
```
Neat!

I like the idea of having a file that represents our custom choices, and then having the main file be generated from applying that one on top of the upstream data sources.

gnprice commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -0,0 +1,141 @@
+                chr(int(h, 16))
+                for h in (emoji_dict["non_qualified"] or emoji_dict["unified"]).split("-")
+            )
+            if emoji not in CLDR_DATA:
```
Can you include a prep commit that just re-sorts the existing `emoji_names.py` (and I guess drops its comments) in the way this script does, before the commit that actually starts using CLDR data? That way, the diff on the latter commit will show just the substantive changes from the new data, which will be helpful for understanding the changes.

I think this can be accomplished by first having a version of this script that doesn't look at CLDR data, and running that. Then the later commit adds the parts of this script that do use CLDR, and the results of running that.

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -1,1580 +1,2055 @@
+    "1f646-200d-2640": {"canonical_name": "woman gesturing OK", "aliases": []},
+    "1f646-200d-2642": {"canonical_name": "man gesturing OK", "aliases": []},
+    "1f646": {"canonical_name": "ok_signal", "aliases": []},
+    "1f647-200d-2640": {"canonical_name": "woman bowing", "aliases": []},
+    "1f647-200d-2642": {"canonical_name": "man bowing", "aliases": []},
+    "1f647": {"canonical_name": "bow", "aliases": []},
```
Hmm. How does it work out when there are spaces in the `canonical_name` like this? Does typeahead give you e.g. `:man bowing:`? Does that get rendered successfully as the intended emoji?

For as long as we have the model that emoji have names and the names appear in message Markdown, I think we probably want to keep the names in the lowercase-underscores style. That should mostly be a matter of replacing `' '` with `'_'`, and perhaps a few other replacements.

evykassirer replied:
We already have spaces! Check out `:stuck out tongue:` .... oh I see, it shows it without underscores in the search view but uses underscores when you complete typing it (and in `emoji_names.py`). I'll play around with fixing that!

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -1,1580 +1,2055 @@
+    "1f1f8-1f1ec": {"canonical_name": "flag: Singapore", "aliases": []},
+    "1f1f8-1f1ed": {"canonical_name": "flag: St. Helena", "aliases": []},
+    "1f1f8-1f1ee": {"canonical_name": "flag: Slovenia", "aliases": []},
+    "1f1f8-1f1ef": {"canonical_name": "flag: Svalbard & Jan Mayen", "aliases": []},
```
Similarly, one might have to delete the colons (I think you mentioned somewhere that searching for flag emojis in particular wasn't working in your testing -- perhaps this is why?).

Might also want to replace these `&` with `and`, and lowercase all the letters.

evykassirer replied:
There are some that work with colons, and I tried removing "flag:" and the flags still didn't work, which leaves me pretty stumped.

Lowercasing and replacing "&" with "and" (in the script) both sound good!

gnprice replied:
Huh. I'll be curious what it turns out to be, then!

evykassirer replied:
It was that there can't be any uppercase letters in the canonical names! The string comparison always turns the search string into lowercase and compares to names/aliases.

gnprice replied:
Ah! That makes sense, then.

So implicitly the search is assuming the names are all lowercase, while it tries to implement a case-insensitive search. It wouldn't be hard to fix it to not make that assumption, but also we probably want these colon-style names to be in all lowercase anyway: https://github.com/zulip/zulip/pull/22067#discussion_r896256356

gnprice replied:
This suggests a possible bug in our handling of custom emoji. We should either (a) enforce that custom-emoji names are in lowercase, or (b) make the search work properly when they're not. It's possible we already do (a).

evykassirer replied:
so... I think we do enforce it, but we don't have good error messages haha.

<img width="738" alt="image" src="https://user-images.githubusercontent.com/5634097/176826922-64623bc9-e719-4b99-8bb5-736e35561d62.png">

evykassirer replied:
https://chat.zulip.org/#narrow/stream/137-feedback/topic/emoji.20validation

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -1,1580 +1,2055 @@
-        "aliases": ["smiling_imp", "smiling_face_with_horns"],
+    "1f21a": {"canonical_name": "japanese_free_of_charge_button", "aliases": ["u7121"]},
+    "1f22f": {"canonical_name": 'Japanese "reserved" button', "aliases": ['"reserved"', "\u6307"]},
+    "1f232": {"canonical_name": "japanese_prohibited_button", "aliases": ["u7981"]},
+    "1f233": {"canonical_name": 'Japanese "vacancy" button', "aliases": ['"vacancy"', "\u7a7a"]},
+    "1f234": {"canonical_name": "japanese_passing_grade_button", "aliases": ["u5408"]},
```
I think rather than override the directional quotes to straight quotes, it might be better to just delete the quotes (and have the script do that.)

evykassirer replied:
I love it when scripts do things. Thanks for the suggestion :)

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -1,1580 +1,2055 @@
+    "1f1ff-1f1e6": {"canonical_name": "flag: South Africa", "aliases": []},
+    "1f1ff-1f1f2": {"canonical_name": "flag: Zambia", "aliases": []},
+    "1f1ff-1f1fc": {"canonical_name": "flag: Zimbabwe", "aliases": []},
+    "1f201": {"canonical_name": 'Japanese "here" button', "aliases": ['"here"\u30b3\u30b3']},
```
I think the corresponding line for this in the exceptions file is missing a comma. An illustration of one reason why it's helpful to have the script take care of these fixes automatically where possible :slightly_smiling_face:

gnprice reviewed (commented):
Exciting! Comments below.

andersk commented on zerver/lib/events.py:
```diff
@@ -356,6 +357,9 @@ def fetch_initial_state_data(
         state["realm_emoji"] = realm.get_emoji()
 
+    if want("emoji_data_url"):
+        state["emoji_data_url"] = staticfiles_storage.url("generated/emoji/emoji_api.json")
```
Note: this needs #22275 for useful `Cache-Control` headers.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9225,6 +9225,10 @@ paths:
+                      emoji_data_url:
+                        type: string
+                        description: |
+                          The URL to a JSON file that describes which emoji names map to which emoji codes.
```
Here's a draft incorporating the other details of this API we want to specify, based on the chat thread:

```suggestion
                          The URL to a JSON file that describes which emoji names map to which emoji codes, for all Unicode emoji this Zulip server accepts.

                          The data at the given URL is a JSON object with one property, `name_to_code`. The value of that property is a JSON object where each key is an [emoji name](/api/add-reaction#parameter-emoji_name) for an available Unicode emoji, and each value is the corresponding [emoji code](/api/add-reaction#parameter-emoji_code). A given emoji code may appear under multiple emoji names.

                          The data at this URL is immutable: the same URL will always represent the same JSON data. This means that if a client sees the value of `emoji_data_url` is a URL it has previously fetched, it can rely on the data it fetched then and avoid fetching it again. (The server accomplishes this guarantee by incorporating a hash of the data in the URL.)
```

(But do wrap the paragraphs -- that'd just be annoying to do here in the GitHub comment editor.)

I think ideally we'd use the nice OpenAPI syntax for describing the structure of the JSON object. Not sure how much fiddling that might require. But I think for now, with just the one property `name_to_code`, this prose version works well enough.

gnprice replied:
Hmm. That is a problem for the mobile app, then -- we don't have an easy way to get our hands on the HTTP headers.

If that's the situation, then I think we should go back to having an `emoji_data_hash` property in the API too.

gnprice replied:
(Subsequent thread on the new PR for just this part of the changes: https://github.com/zulip/zulip/pull/22476#discussion_r923967310 )

gnprice commented on tools/setup/emoji/build_emoji:
```diff
@@ -349,6 +383,22 @@ def generate_map_files(cache_path: str, emoji_catalog: Dict[str, List[str]]) ->
+        emoji_api_file.write(
+            orjson.dumps(
+                {
+                    "name_to_codepoint": name_to_codepoint,
```
Should be `name_to_code` (or some other name we pick):
https://chat.zulip.org/#narrow/stream/378-api-design/topic/currently.20supported.20emoji/near/1385346

(`name_to_codepoint` is an inaccurate name, because each value can represent a whole sequence of codepoints, not just one. Plus, they're encoded in [an idiosyncratic format](https://zulip.com/api/add-reaction#parameter-reaction_type):
> `unicode_emoji` : In this namespace, `emoji_code` will be a dash-separated hex encoding of the sequence of Unicode codepoints that define this emoji in the Unicode specification.

so the only way to really make sense of them is by reference to the `emoji_code` concept in the Zulip API.)

evykassirer replied:
I copied this from `emoji_codes.json` using the same key for the same data, but that makes sense!

gnprice replied:
Yep, this is specifically something I wanted to fix relative to what's in that file :slightly_smiling_face:

gnprice reviewed (commented):
Thanks @evykassirer ! Comments below on the newest commit, adding the emoji-data API.

I believe this now fixes #21037, so the PR description should be updated to say that (so that merging it will close the issue.)

Now that there's this server-side implementation of this new API feature, we can implement the mobile side -- https://github.com/zulip/zulip-mobile/issues/2956 -- and test it against a dev server with this PR. I've just marked that as unblocked, and we'll aim to do it soon, which will be good for getting end-to-end feedback on the API change.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9225,6 +9225,24 @@ paths:
+                        type: string
+                        description: |
+                          The URL to a JSON file that describes which emoji names map to which
+                          emoji codes, for all Unicode emoji this Zulip server accepts. The data
+                          at the given URL is a JSON object with one property, `name_to_code`.
```
Paragraph break here:

```suggestion
                          emoji codes, for all Unicode emoji this Zulip server accepts.

                          The data
                          at the given URL is a JSON object with one property, `name_to_code`.
```

Also before "The data at this URL is immutable", making a third paragraph. (Though based on the discussion up to here:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/currently.20supported.20emoji/near/1397835
the current plan is we'll actually just delete that third paragraph.)

gnprice reviewed (commented):
Thanks! Just one comment on the revised version.

timabbott commented on tools/setup/emoji/build_emoji:
```diff
@@ -384,6 +384,22 @@ def generate_map_files(cache_path: str, emoji_catalog: Dict[str, List[str]]) ->
+                    "name_to_code": name_to_codepoint,
+                }
+            )
+        )
```
Is it easy to move this commit to a separate PR? I'll post a few other comments on some API changelog details with this, but I think it can be merged independently/before the rest of these changes, and it'll simplify things to do that.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9225,6 +9225,18 @@ paths:
+                          [emoji name](/api/add-reaction#parameter-emoji_name) for an available
+                          Unicode emoji, and each value is the corresponding
+                          [emoji code](/api/add-reaction#parameter-emoji_code).
+                          A given emoji code may appear under multiple emoji names.
```
This API change needs to be documented with a `**Changes**` entry and API_FEATURE_LEVEL bump (see `version.py` for documentation).

timabbott commented:
Can you adjust the commit messages for the first few commits to explain to what extent they contain any function/user-facing changes? I think my read is that at least the first two commits are intended to have no functional changes yet, but I'm not sure, and I'm a bit confused that the second commit adds the CLDR dependency, but only the 3rd commit discusses using CLDR data. 

(My goal here is to be able to merge the first couple commits, since they seem like they might be lower risk than the cutover)

timabbott commented:
(I guess to be explicit, I think the strategy of extracting a `custom_emoji_names.py` file is a good one).

evykassirer commented:
Also thanks for pointing out the CLDR package thing, it was indeed in the wrong commit!

evykassirer commented:
#22475 #22476

alya commented:
No feedback from my end after playing with this a bit. :)

timabbott commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -69,6 +69,23 @@ SKIN_TONE_EMOJI_CODES = [
+    }
+    for before, after in replacements.items():
+        name = name.replace(before, after)
+    return name.lower()
```
I think this algorithm might be incomplete; this commit leaves us with a `flag R\u00e9union` canonical name.

evykassirer replied:
This was handled in the custom emoji list [here](https://github.com/zulip/zulip/pull/22067/files/7a18bbe555e2598e9cffa847f52914b3574b2ece#diff-bdbbd354daa882573f5d612653e8e18dc4791a25723fe1fa58fdc7dd313aac18R1595). (I forgot to put `_` there, which signals to me that it would be good to run `cleanup` on the custom names as well as the CLDR names just in case. )

I kept the correct-in-that-language spelling as the canonical name in situations like that (since I felt like that was respectful to that language), and added aliases for "English"-letter spellings so that the emoji could still be accessible without having to type those characters. I can add a comment explaining that thinking. 

I hadn't generalized that because there are a lot of non-standard characters, but I just added `convert_non_ascii_chars` and a block of code that uses that. Let me know what you think :)

timabbott replied:
I think this is a reasonable policy -- but this sort of decision needs to be documented with comments :)

evykassirer replied:
I had used a short comment elsewhere in the code that was a TODO describing that I wasn't doing this yet. Agreed that I could have added more context about why I didn't, and maybe also made a comment in the `cleanup` function. But there definitely were comments about it.

Now there is a function that handles this, so it feels like I don't need to comment anymore, especially because an error will be raised if any new non-ascii characters are added in the future.

timabbott commented on static/js/tippyjs.js:
```diff
@@ -210,6 +210,14 @@ export function initialize() {
+        onHidden(instance) {
+            instance.destroy();
+        },
+    });
```
Can you add a comment documenting why this one needs custom behavior?

evykassirer replied:
At this point I don't remember why I put it there, and it seems like things work fine without it so I will take it out. 🤷 thanks!

timabbott commented on static/templates/settings/display_settings.hbs:
```diff
@@ -59,6 +59,14 @@
+                        <span>(<a href='/help/emoji-and-emoticons#change-your-emoji-set'>deprecated</a>)</span>
+                        <i
+                          class="fa fa-question-circle-o tippy-zulip-tooltip"
+                          data-tippy-placement="bottom"
```
Let's move the placement into the `tippyjs.js` file -- if you're going to have custom styling for its tooltip anyway, best to have it all in one place.

timabbott commented on zerver/migrations/0407_rename_emoji_types.py:
```diff
@@ -0,0 +1,41 @@
+                max_length=20,
+            ),
+        ),
+    ]
```
Since this change is just changing the Django metadata about (which doesn't appear in the database), it's better to delete this migration file and instead just reorder the `choices` in the original migration -- saves a migration for everyone, which will be nice for folks testing this PR.

evykassirer replied:
Oh neat I didn't know I could do that! Thanks

timabbott commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -72,9 +108,47 @@ def main() -> None:
+                    f"{emoji} not found in custom emoji name maps, but also not found in CLDR data. Skipping."
+                )
+                continue
+            assert len(CLDR_DATA[emoji]["tts"]) == 1
```
Can you add a comment for what this assertion is doing? I have no idea what `tts` means. 

Another idea that might be better is to add a comment with a link to the part of the documentation for the CLDR data library defining its format; or if that doesn't exist, just give an example shape in a comment.

Would make this a lot easier to read :)

andersk replied:
The documentation is in [Unicode Technical Standard #35](https://www.unicode.org/reports/tr35/tr35-general.html#14-annotations-and-labels).

timabbott replied:
Cool, let's add a link to that standard in the comment as well.

timabbott commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -72,9 +108,47 @@ def main() -> None:
+        if len(alias_to_emoji_code[alias]) > 1:
+            for emoji_code in alias_to_emoji_code[alias]:
+                if emoji_code not in CUSTOM_EMOJI_NAME_MAPS:
+                    all_emojis[emoji_code]["aliases"].remove(alias)
```
Is the intent that we'd resolve this sort of conflict via `CUSTOM_EMOJI_NAME_MAPS` in general? If they're uncommon, I'm kinda wondering if it'd make sense to have an assert statement here instead.

evykassirer replied:
I can add a comment here for why we needed to add this, but essentially the CLDR data doesn't have the same sense of unique alias that we do, so we have to filter it out ourselves. I don't want to hardcode it all in because we should refer to CLDR data as a point of truth as much as possible. If the alias was specifically specified in the custom list then we can keep just that one, but otherwise I just removed all of them instead of randomly picking a favourite that gets to use it. Does that make sense?

timabbott replied:
The comment here is very helpful.

timabbott commented:
Thanks for all the work on this @evykassirer! I think we should be able to deploy this to chat.zulip.org soon, but I'm going to wait until we've resolved the above comments since I don't fully understand some of the code.

evykassirer commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -53,17 +69,61 @@ SKIN_TONE_EMOJI_CODES = [
+        "ó": "o",
+        "ā": "a",
+        "ō": "o",
+        "&": "and",
```
oops this shouldn't be here

andersk commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -32,8 +34,279 @@
+    "1f1e6-1f1f9": {"canonical_name": "flag_austria", "aliases": []},
+    "1f1e6-1f1fa": {"canonical_name": "flag_australia", "aliases": []},
+    "1f1e6-1f1fc": {"canonical_name": "flag_aruba", "aliases": []},
+    "1f1e6-1f1fd": {"canonical_name": "flag_\u00e5land_islands", "aliases": ["flag_aland_islands"]},
```
This file would be more readable without the `\u` escapes, which is easy to fix:

```diff
--- a/tools/setup/emoji/generate_emoji_names
+++ b/tools/setup/emoji/generate_emoji_names
@@ -82,7 +82,7 @@ def main() -> None:
             "EMOJI_NAME_MAPS: Dict[str, Dict[str, Any]] = {\n"
         )
         for (key, emoji_names) in all_emojis.items():
-            f.write(f'    "{key}": {json.dumps(emoji_names)},\n')
+            f.write(f"    {key!r}: {emoji_names!r},\n")
         f.write("}\n")
 
     print(
```

evykassirer replied:
great, thanks!

timabbott commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -868,26 +1361,54 @@
     "1f926": {"canonical_name": "face_palm", "aliases": []},
     "1f927": {"canonical_name": "sneezing", "aliases": []},
+    "1f928": {"canonical_name": "face_with_raised_eyebrow", "aliases": ["distrust", "skeptic"]},
+    "1f929": {"canonical_name": "star-struck", "aliases": []},
```
I think we might treat `-`, ` `, and `_` as equivalent in our matching algorithm implementation -- should this be `star_struck`? I don't think we should intentionally have any `-`s in the `canonical_name` field.

rht replied:
It's `star-struck` on Slack and Matrix (both using iamcal). It would be harder to bridge the reactions between the platforms if the canonical name is manually set to `star_struck`.

andersk replied:
Any code that assumes emoji are represented by names is already wrong, but any code that assumes emoji are represented by the *same* name on *three* wrong platforms is *quadruply* wrong. Colon strings are not a stable interchange format.

rht replied:
The colon strings in Slack and Matrix are unlikely to change because they have been used in production for a while, and changing them would break compatibility. But I suppose my concern is easily solvable by using `zerver.lib.emoji.name_to_codepoint` in the bridge.

But then again, having the new colon strings that deviate from the most common standards would make it harder to bridge when adding emoji support in Matterbridge. To bridge reactions between Slack and Matrix, it doesn't require any mapping of name -> codepoint -> name.

andersk replied:
Emoji are Unicode characters. Literally the whole point of Unicode is to avoid these kind of stupid interchange problems. There is no good reason to map a codepoint that works everywhere to a name that’s only meaningful to some specific applications.

rht replied:
It's not my choice that Slack and Mattermost (not sure about Matrix) have colon string reactions in their message JSON. As such, it is necessary to know what naming scheme they use anyway.

timabbott replied:
Any code that importing/mirroring content between Zulip and third-party services will need to do something to map names for emoji, because every service uses their own set of names. It shouldn't be too hard to do name=>codepoint=>other name translations (or probably better in most cases, name => codepoint translations, i.e. just replacing `:star-struck:` in a slack message with the 🤩 unicode character) as needed in tools whose purpose is to translate between different protocols. 

However, we still need to answer my original question: Does `star-struck` and anything else with a `-` in it need to be spelled `star_struck` in order to not be broken/unusable in Zulip's emoji system?

evykassirer replied:
I think it was usable before, but I switched it to be all underscores for consistency.

gnprice commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -72,17 +131,78 @@ def main() -> None:
+            if not name.isascii():
+                ascii_alias = convert_non_ascii_chars(name)
+                # Now no other emoji can use this alias.
+                for code in alias_to_emoji_code[ascii_alias]:
+                    all_emojis[code]["aliases"].remove(alias)
```
Is this meant to say `.remove(ascii_alias)`, matching the iteration? If not, then the mismatch deserves a comment explaining the reasoning for what's going on.

evykassirer replied:
Yup this is a mistake, thanks for catching it!

gnprice commented on tools/setup/emoji/emoji_names.py:
```diff
@@ -32,8 +34,279 @@
+    "1f1e7-1f1ff": {"canonical_name": "flag_belize", "aliases": []},
+    "1f1e8-1f1e6": {"canonical_name": "flag_canada", "aliases": []},
+    "1f1e8-1f1e8": {"canonical_name": "flag_cocos_(keeling)_islands", "aliases": []},
+    "1f1e8-1f1e9": {"canonical_name": "flag_congo___kinshasa", "aliases": []},
+    "1f1e8-1f1eb": {"canonical_name": "flag_central_african_republic", "aliases": []},
+    "1f1e8-1f1ec": {"canonical_name": "flag_congo___brazzaville", "aliases": []},
```
The `___` in these two names would be good to shorten. A line at the end of `cleanup_name`, replacing the regexp `_{2,}` (i.e., a run of two or more `_` characters) with the string `_`, would do it.

gnprice commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -72,17 +131,78 @@ def main() -> None:
+            else:
+                alias_to_emoji_code[alias].append(emoji_code)  # This is used in STEP 3.
+
+    # STEP 3: We don't support having aliases for multiple emoji, but the CLDR data doesn't
+    # have that same restriction, so we have to fix this up to have unique aliases.
```
This wording is a bit confusing -- I think I'd have trouble working out what "having aliases for multiple emoji" means if I didn't already know what's going on. (After all, lots of emoji -- multiple of them, for sure -- do have aliases, and that's fine.)

I think it can be cleared up basically by using fewer plurals: we don't support multiple emoji sharing the same alias.

gnprice reviewed (commented):
Thanks @evykassirer! Happy to see this approaching merge. Small comments below.

evykassirer commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -72,17 +131,78 @@ def main() -> None:
+    # way to spell that word, but always add an alias for an ascii-only version of the word.
+    for (emoji_code, emoji_names) in all_emojis.items():
+        for name in [emoji_names["canonical_name"]] + emoji_names["aliases"]:
+            # These are known names where we don't have an ascii-only version and there are ascii aliases
+            # that a user can still enter instead to get the same emoji.
+            if name in ["ココ", "サ", "指", "空"]:
+                assert [alias.isascii() for alias in aliases]
```
oh whoops I definitely meant to do `[alias for alias in aliases if alias.isascii()]` 😅 

I like the `any` syntax better though

evykassirer commented:
Thanks for the great comments Greg! Just made those changes.

timabbott commented:
A commit message nit: The commit title for the /help/ edit should something like `help: Document Google blob deprecation.` -- we use `help:` as the prefix for that (`docs:` is reserved for ReadTheDocs content, aka `docs/` in the codebase). Though actually ... at the least the part of that commit that changes the labels would ideally be part of the previous commit, which actually changes those labels.

In any case, I'm test-deploying this on chat.zulip.org!

timabbott commented:
The chat.zulip.org deployment seems to be running happily without incident, so we should merge this soon.

One bug I've noticed is we're missing `target="_blank`` (and the `rel=` tags required for security) in the /help/ link about the deprecation; this means that clicking that links opens the /help/ article in the same window.

I believe our plans included a small migration to reset everyone using the Google blobs emoji to the modern Google emojiset; can you add a migration for that? We didn't want that on chat.zulip.org so we'd test both emoji set choices, but probably do for what we merge.

timabbott commented:
@alya @laurynmm I think it's reasonable for you to do a final review on the /help/ content for this change.

alya commented:
I think the help text looks fine.

Poking around the setting though, I have a couple of thoughts:

1. Can we drop the help center link on `deprecated`. Having actually experienced it, it feels confusing, and the information in the help center is roughly the same as what's explained in the `?` tooltip on the same line.
2. I wonder if we should italicize `deprecated`? I think that would match better the formatting style we use for other settings.

Thanks!

alya commented:
We should also think about alternative ways to explain what's up with these emoji being deprecated.  I think this is the only place where the `?` reveals a tooltip rather than linking to the help center, and it's not great to break that pattern.

We could have the `?` link to the help center, but that does make the relevant text significantly harder to find.

We could also put the text directly into the settings page rather than hiding it in a tooltip, e.g. by displaying it only if "Google blobs" is the selected option.

laurynmm commented:
> I think this is the only place where the `?` reveals a tooltip rather than linking to the help center, and it's not great to break that pattern.

I think we use the "fa fa-question-circle" when it's a tooltip (see `@all` note in notifications settings table) and "fa fa-question-circle-o" when it's a help center link.

> We could have the `?` link to the help center, but that does make the relevant text significantly harder to find.

If we link directly to the **Change your emoji set** section, then it wouldn't be too hard to find, which is what the current link in "deprecated" does.

---

My only other documentation note would be to maybe put the help center text in a [warning note](https://zulip.readthedocs.io/en/latest/documentation/helpcenter.html#tips-and-warnings), but that might draw more attention to it than we want? See screenshot below.

<details>
<summary>Google blob note as warning screenshot</summary>

![Screenshot from 2022-09-28 12-38-59](https://user-images.githubusercontent.com/63245456/192759053-f499419c-de63-446d-8680-cc4abc3002a4.png)
</details>

alya commented:
> I think we use the "fa fa-question-circle" when it's a tooltip (see `@all` note in notifications settings table) and "fa fa-question-circle-o" when it's a help center link.

Oh, I forgot about that one! (We've been removing `i` icons, which previously offered tooltips in other parts of the menus, in favor of other approaches.) 

I'm a bit confused -- the `?` on all visually looks like "fa fa-question-circle"?

<img width="68" alt="Screen Shot 2022-09-29 at 1 49 16 PM" src="https://user-images.githubusercontent.com/2090066/193139570-4cc9944e-954f-4c6d-8184-5441d21258b0.png">

If it is somehow different, I'm not sure that would be intentional.

alya commented:
Anyway, what if we combine the ideas above and use the warning styling @laurynmm suggested for the help center, and have the question mark link to that help center section? That seems clear enough?

evykassirer commented:
updates:

* updated commit message to "help: Document Google blob deprecation"
* moved link on "deprecated" to the icon and removed tooltip
* added warning styling to helpcenter
* added the new migration and tested it locally

<img width="405" alt="image" src="https://user-images.githubusercontent.com/5634097/193654064-760cb2bf-2df2-4132-9af3-ce8c0ff5cb29.png">

<img width="835" alt="image" src="https://user-images.githubusercontent.com/5634097/193655282-1c909e51-c837-4667-9543-b828f240e06e.png">

evykassirer commented:
(I reverted the migration locally and tried again and it didn't seem to work, so I'll look into it more later)

evykassirer commented:
Okay I feel more confident now having written and run this test successfully:

```py
class TempTestForEmojiMigration(MigrationsTestCase):
    migrate_from = "0415_delete_scimclient"
    migrate_to = "0416_set_default_emoji_style"

    @use_db_models
    def setUpBeforeMigration(self, apps: StateApps) -> None:
        iago = self.example_user("iago")
        iago.emojiset = "google-blob"
        iago.save()
        iago = self.example_user("iago")
        self.assertEqual(iago.emojiset, "google-blob")

    def test_emoji_updates(self) -> None:
        iago = self.example_user("iago")
        self.assertEqual(iago.emojiset, "google")


class TempTest2ForEmojiMigration(MigrationsTestCase):
    migrate_from = "0415_delete_scimclient"
    migrate_to = "0416_set_default_emoji_style"

    @use_db_models
    def setUpBeforeMigration(self, apps: StateApps) -> None:
        RealmUserDefault = apps.get_model("zerver", "RealmUserDefault")
        Realm = apps.get_model("zerver", "Realm")
        realm = Realm.objects.get(string_id="zulip")
        realm_user_default = RealmUserDefault.objects.get(realm=realm)
        realm_user_default.emojiset = "google-blob"
        realm_user_default.save()

        iago = self.example_user("iago")
        iago.emojiset = "google-blob"
        iago.save()
        iago = self.example_user("iago")
        self.assertEqual(iago.emojiset, "google-blob")

    def test_emoji_updates2(self) -> None:
        iago = self.example_user("iago")
        self.assertEqual(iago.emojiset, "google-blob")
```

timabbott commented on zerver/migrations/0416_set_default_emoji_style.py:
```diff
@@ -0,0 +1,30 @@
+    ]
+
+    operations = [
+        migrations.RunPython(update_deprecated_emoji_style),
```
This should be marked `elidable=True`, since this migration is a noop for an empty database.

timabbott commented on static/templates/settings/display_settings.hbs:
```diff
@@ -59,6 +59,12 @@
                             {{/if}}
                         </span>
+                        {{#if (eq this.key "google-blob")}}
+                        <span><em>(deprecated)</em></span>
```
"deprecated", like any other user facing text, needs to be tagged for translation. I also moved the `(` outside the `em` tag, to match how our other settings with this pattern do it.

timabbott commented:
I applied a few final changes, noted above, and marked this to merge once CI finishes. Thanks for making this happen @evykassirer!

rht commented:
What is the reason for "1f929" to have "star_struck" as its canonical name? The CLDR name is "grinning_face_with_star_eyes". Shouldn't "star_struck" be an alias instead?

andersk commented:
@rht No, GRINNING FACE WITH STAR EYES is the Unicode character name, not the CLDR name. These are [not the same thing](https://chat.zulip.org/#narrow/stream/101-design/topic/emoji.20names/near/1381335).

https://github.com/unicode-org/cldr/blob/release-41/common/annotations/en.xml#L827

rht commented:
OK that clarifies it, thank you!

rht commented:
I was assuming that Slack (which has both "star-struck" and "grinning_face_with_star_eyes" for "1f929") takes it from CLDR. It turns out to be not the case.

andersk commented on tools/setup/emoji/generate_emoji_names:
```diff
@@ -72,17 +133,78 @@ def main() -> None:
+            # These are known names where we don't have an ascii-only version and there are ascii aliases
+            # that a user can still enter instead to get the same emoji.
+            if name in ["ココ", "サ", "指", "空"]:
+                assert any(alias.isascii() for alias in aliases)
```
You meant `emoji_names["aliases"]`, right?

- #26465
